processing/src/main/java/io/druid/query/filter/SelectorDimFilter.java

```diff
@@ -51,25 +63,57 @@ public SelectorDimFilter(
   {
     byte[] dimensionBytes = StringUtils.toUtf8(dimension);
     byte[] valueBytes = (value == null) ? new byte[]{} : StringUtils.toUtf8(value);
+    byte[] extractionFnBytes = extractionFn == null ? new byte[0] : extractionFn.getCacheKey();
 
-    return ByteBuffer.allocate(2 + dimensionBytes.length + valueBytes.length)
+    return ByteBuffer.allocate(3 + dimensionBytes.length + valueBytes.length + extractionFnBytes.length)
                      .put(DimFilterCacheHelper.SELECTOR_CACHE_ID)
                      .put(dimensionBytes)
                      .put(DimFilterCacheHelper.STRING_SEPARATOR)
                      .put(valueBytes)
+                     .put(DimFilterCacheHelper.STRING_SEPARATOR)
+                     .put(extractionFnBytes)
                      .array();
   }
 
   @Override
   public DimFilter optimize()
   {
+    if (this.getExtractionFn() instanceof LookupExtractionFn
+        && ((LookupExtractionFn) this.getExtractionFn()).isOptimize()) {
+      LookupExtractionFn exFn = (LookupExtractionFn) this.getExtractionFn();
+
+      // We cannot do an unapply()-based optimization if the selector value
+      // and the replaceMissingValuesWith value are the same, since we have to match on
+      // all values that are not present in the lookup.
+      if (Objects.equals(Strings.emptyToNull(this.getValue()), exFn.getReplaceMissingValueWith())) {
+        return this;
+      }
+
+      LookupExtractor lookup = exFn.getLookup();
+      final List<String> keys = lookup.unapply(this.getValue());
+      final String dimensionName = this.getDimension();
+      if (!keys.isEmpty()) {
+        // If retainMissingValues is true, there may be row values that match the selector value
+        // but are not included in the lookup map. Match on the selector value as well.
+        if (exFn.isRetainMissingValue()) {
+          keys.add(value);
+        }
+
+        return new InDimFilter(dimensionName, keys, null);
+      } else {
+        // Nothing in the lookup would match, just use a plain selector filter.
+        if (exFn.isRetainMissingValue()) {
+          return new SelectorDimFilter(dimension, value, null);
```

Comment:
I think this would be clearer if the `else` and `if` were combined into an `else if`. It made me do a double-take originally.

Would be even clearer to include an `else { return this }` too, just to cover all the cases explicitly.